pkg/sdkserver/run.go

```diff
@@ -0,0 +1,172 @@
+package sdkserver
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"github.com/gptscript-ai/gptscript/pkg/gptscript"
+	"github.com/gptscript-ai/gptscript/pkg/loader"
+	"github.com/gptscript-ai/gptscript/pkg/mvl"
+	"github.com/gptscript-ai/gptscript/pkg/runner"
+	gserver "github.com/gptscript-ai/gptscript/pkg/server"
+	"github.com/gptscript-ai/gptscript/pkg/types"
+)
+
+type loaderFunc func(context.Context, string, string, ...loader.Options) (types.Program, error)
+
+func (s *server) execAndStream(ctx context.Context, programLoader loaderFunc, logger mvl.Logger, w http.ResponseWriter, opts *gptscript.Options, chatState, input, subTool string, toolDef fmt.Stringer) {
+	g, err := gptscript.New(opts)
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to initialize gptscript: %w", err))
+		return
+	}
+	defer g.Close(false)
+
+	prg, err := programLoader(ctx, toolDef.String(), subTool, loader.Options{Cache: g.Cache})
+	if err != nil {
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to load program: %w", err))
+		return
+	}
+
+	errChan := make(chan error)
+	programOutput := make(chan runner.ChatResponse)
+	events := s.events.Subscribe()
+	defer events.Close()
+
+	go func() {
+		run, err := g.Chat(ctx, chatState, prg, opts.Env, input)
+		if err != nil {
+			errChan <- err
+		} else {
+			programOutput <- run
+		}
+		close(errChan)
+		close(programOutput)
+	}()
+
+	processEventStreamOutput(ctx, logger, w, gserver.IDFromContext(ctx), events.C, programOutput, errChan)
+}
+
+// processEventStreamOutput will stream the events of the tool to the response as server sent events.
+// If an error occurs, then an event with the error will also be sent.
+func processEventStreamOutput(ctx context.Context, logger mvl.Logger, w http.ResponseWriter, id string, events <-chan gserver.Event, output <-chan runner.ChatResponse, errChan chan error) {
+	run := newRun(id)
+	setStreamingHeaders(w)
+
+	streamEvents(ctx, logger, w, run, events)
+
+	var out runner.ChatResponse
+	select {
+	case <-ctx.Done():
+	case out = <-output:
+		run.processStdout(out)
+
+		writeServerSentEvent(logger, w, map[string]any{
+			"stdout": out,
+		})
+	case err := <-errChan:
+		writeError(logger, w, http.StatusInternalServerError, fmt.Errorf("failed to run file: %w", err))
+	}
+
+	// Now that we have received all events, send the DONE event.
+	_, err := w.Write([]byte("data: [DONE]\n\n"))
+	if err == nil {
+		if f, ok := w.(http.Flusher); ok {
+			f.Flush()
+		}
+	}
+
+	logger.Debugf("wrote DONE event")
+}
+
+// streamEvents will stream the events of the tool to the response as server sent events.
+// This looks for and tries to handle confirm events as well. However, that currently is not implemented in the SDK.
```

Comment:
What's the confirm event below?

Reply:
This changed from the initial implementation. I removed the second line of this comment.